OutOfSchool/OutOfSchool.Redis/CacheService.cs

```diff
@@ -95,24 +93,71 @@ await ExecuteRedisMethod(() =>
     }
 
     public Task RemoveAsync(string key)
-        => ExecuteRedisMethod(async () => {
+        => this.ExecuteRedisMethod(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                cache.Remove(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+
+    public async Task<string> GetValueAsync(string key)
+    {
+        string returnValue = string.Empty;
+
+        await ExecuteRedisMethod(() =>
+        {
+            cacheLock.EnterReadLock();
+            try
+            {
+                returnValue = cache.GetString(key);
+            }
+            finally
+            {
+                cacheLock.ExitReadLock();
+            }
+        });
+
+        return returnValue;
+    }
+
+    public async Task SetValueAsync<T>(
+        string key,
+        T value,
+        TimeSpan? absoluteExpirationRelativeToNowInterval = null,
+        TimeSpan? slidingExpirationInterval = null)
+    {
+        await ExecuteRedisMethod(() =>
+        {
```

Comment:
(Code duplication) This method looks like a copy of part of the `GetOrAdd()` implementation. 
If it is so and to avoid duplication could it be called in `GetOrAdd()` instead of existing code?
If it is not a duplication - then why? For me it seems to do same thing

Reply:
Today, I've had a discussion about MementoService and CacheService with @h4wk13.
He suggested not to create a new entity (RequiredWorkshopMemento) and then serialize it and pass it to Cache. Instead, work directly with the Json-string and store it into the cache. Then all work will be done in CacheService class.
I asked this question to @DmyMi and now I'm wating the answer.

Reply:
> (Code duplication) This method looks like a copy of part of the `GetOrAdd()` implementation. If it is so and to avoid duplication could it be called in `GetOrAdd()` instead of existing code? If it is not a duplication - then why? For me it seems to do same thing

I agree with you that the code in GetOrAddAsync<T>() method is similar to the code in new methods.
In my opinion, this happened because this method should not be related to the ICacheService interface, since it has a query to the database. 
I think that a new interface should be created (I don’t know what to call it yet) in which the GetOrAddAsync<T>() method will be included. And all other methods (RemoveAsync(), GetValueAsync() and UpsertValueAsync<T>()) have to be in ICasheService interface. Then ICrudCacheService does not make sense and should be removed.

Reply:
Sorry, where we have query to db in this code?

Reply:
parameter - Func<Task<T>> newValueFactory,
if (!isExists)
{
    returnValue = await newValueFactory();
...

Reply:
I get it finally. It's not a "query to db" its just a "value factory". And I see the difference now.
But I still don't see why it's impossible to reduce code duplication. You can use `Upsert` inside `GetOrAdd`.

Reply:
Ok, I used ReadAsync() and WriteAsync inside GetOrAdd.